addon/-private/functional/modifier.ts

```diff
@@ -54,14 +60,72 @@ type Teardown = () => unknown;
 export default function modifier<
   E extends Element,
   P extends unknown[],
-  N extends Record<string, unknown>
+  N extends object
 >(
   fn: (element: E, positional: P, named: N) => void | Teardown
 ): FunctionBasedModifier<{
   Args: { Named: N; Positional: P };
   Element: E;
 }>;
 
+/**
+ * An API for writing simple modifiers.
+ *
+ * This function runs the first time when the element the modifier was applied
+ * to is inserted into the DOM, and it *autotracks* while running. Any values
+ * that it accesses will be tracked, including the arguments it receives, and if
+ * any of them changes, the function will run again.
+ *
+ * The modifier can also optionally return a *destructor*. The destructor
+ * function will be run just before the next update, and when the element is
+ * being removed entirely. It should generally clean up the changes that the
+ * modifier made in the first place.
+ *
+ * @param fn The function which defines the modifier.
+ */
+// This overload allows users to write types directly on the callback passed to
+// the `modifier` function and infer the resulting type correctly.
+export default function modifier<
+  E extends Element,
+  P extends unknown[],
+  N extends object
+>(
+  fn: (element: E, positional: P, named: N) => void | Teardown,
+  options: { eager: true }
+): FunctionBasedModifier<{
+  Args: { Named: N; Positional: P };
+  Element: E;
+}>;
+
+/**
+ * An API for writing simple modifiers.
+ *
+ * This function runs the first time when the element the modifier was applied
+ * to is inserted into the DOM, and it *autotracks* while running. Any values
+ * that it accesses will be tracked, including the arguments it receives, and if
+ * any of them changes, the function will run again.
```

Comment:
Maybe reword slightly to ensure that it's clear that _only_ arguments that are "consumed" are included in the "will run again" part.